[Task] Archives incorrectly migrating

To recreate:

1. Git checkout cryostat/cryostat-v2.1
2. `$ podman pull quay.io/cryostat/cryostat:2.1.0 # clear cache! `
3. `$ CRYOSTAT_IMAGE="quay.io/cryostat/cryostat:2.1.0" sh smoketest.sh`
4. create and archive a recording
5. Go back to current commit, and there will be an IOException and the recording is not migrated correctly

I think the migration is ignoring the connectUrl files.

Hmm... in theory it should work, but connecting to the target for the jvmId seems to break and result in archives becoming available for some reason.

```
WARNING: Could not get jvmId for target service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi
Oct 19, 2022 9:56:46 PM io.cryostat.core.log.Logger warn
WARNING: Exception thrown
io.cryostat.recordings.JvmIdHelper$JvmIdGetException: java.util.concurrent.TimeoutException
	at io.cryostat.recordings.JvmIdHelper.getJvmId(JvmIdHelper.java:137)
	at io.cryostat.recordings.RecordingArchiveHelper.lambda$migrate$0(RecordingArchiveHelper.java:177)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1428)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinTask.awaitDone(ForkJoinTask.java:436)
	at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:979)
	at io.cryostat.recordings.RecordingArchiveHelper.migrate(RecordingArchiveHelper.java:214)
	at io.cryostat.recordings.RecordingMetadataManager.lambda$start$7(RecordingMetadataManager.java:389)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1960)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2095)
	at io.cryostat.recordings.JvmIdHelper.getJvmId(JvmIdHelper.java:134)
	... 8 more

Oct 19, 2022 9:56:46 PM io.cryostat.core.log.Logger info
INFO: Successfully migrated archives
```

--- a/src/main/java/io/cryostat/recordings/JvmIdHelper.java
+++ b/src/main/java/io/cryostat/recordings/JvmIdHelper.java
@@ -38,6 +38,7 @@
 package io.cryostat.recordings;
 
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executor;
@@ -58,12 +59,14 @@ import com.github.benmanes.caffeine.cache.AsyncCacheLoader;
 import com.github.benmanes.caffeine.cache.AsyncLoadingCache;
 import com.github.benmanes.caffeine.cache.Caffeine;
 import com.github.benmanes.caffeine.cache.Scheduler;
+import org.apache.commons.codec.binary.Base32;
 
 public class JvmIdHelper extends AbstractEventEmitter<JvmIdHelper.IdEvent, String> {
 
     private final TargetConnectionManager targetConnectionManager;
     private final CredentialsManager credentialsManager;
     private final long connectionTimeoutSeconds;
+    private final Base32 base32;
     private final Logger logger;
 
     private final AsyncLoadingCache<String, String> ids;
@@ -75,10 +78,12 @@ public class JvmIdHelper extends AbstractEventEmitter<JvmIdHelper.IdEvent, Strin
             long connectionTimeoutSeconds,
             Executor executor,
             Scheduler scheduler,
+            Base32 base32,
             Logger logger) {
         this.targetConnectionManager = targetConnectionManager;
         this.credentialsManager = credentialsManager;
         this.connectionTimeoutSeconds = connectionTimeoutSeconds;
+        this.base32 = base32;
         this.logger = logger;
         this.ids =
                 Caffeine.newBuilder()
@@ -109,6 +114,9 @@ public class JvmIdHelper extends AbstractEventEmitter<JvmIdHelper.IdEvent, Strin
                 || targetId.equals(RecordingArchiveHelper.UPLOADED_RECORDINGS_SUBDIRECTORY)) {
             return CompletableFuture.completedFuture(
                     RecordingArchiveHelper.UPLOADED_RECORDINGS_SUBDIRECTORY);
+        } else if (targetId.equals(RecordingArchiveHelper.LOST_RECORDINGS_SUBDIRECTORY)) {
+            return CompletableFuture.completedFuture(
+                    RecordingArchiveHelper.LOST_RECORDINGS_SUBDIRECTORY);
         }
         CompletableFuture<String> future =
                 this.targetConnectionManager.executeConnectedTaskAsync(
@@ -136,6 +144,29 @@ public class JvmIdHelper extends AbstractEventEmitter<JvmIdHelper.IdEvent, Strin
             logger.warn("Could not get jvmId for target {}", targetId);
             throw new JvmIdGetException(e, targetId);
         }
+    }
+
+    public String subdirectoryNameToJvmId(String subdirectoryName) {
+        if (subdirectoryName.equals(RecordingArchiveHelper.UPLOADED_RECORDINGS_SUBDIRECTORY)
+                || subdirectoryName.equals(RecordingArchiveHelper.LOST_RECORDINGS_SUBDIRECTORY)) {
+            return subdirectoryName;
+        }
+        return new String(base32.decode(subdirectoryName), StandardCharsets.UTF_8);
+    }
+
+    public String jvmIdToSubdirectoryName(String jvmId) {
+        if (jvmId.equals(RecordingArchiveHelper.UPLOADED_RECORDINGS_SUBDIRECTORY)
+                || jvmId.equals(RecordingArchiveHelper.LOST_RECORDINGS_SUBDIRECTORY)) {
+            return jvmId;
+        }
+        return base32.encodeAsString(jvmId.getBytes(StandardCharsets.UTF_8));
+    }
+
+    // FIXME: refactor structure to remove file-uploads (v1 RecordingsPostBodyHandler)
+    public boolean isSpecialDirectory(String directoryName) {
+        return directoryName.equals(RecordingArchiveHelper.UPLOADED_RECORDINGS_SUBDIRECTORY)
+                || directoryName.equals("file-uploads")
+                || directoryName.equals(RecordingArchiveHelper.LOST_RECORDINGS_SUBDIRECTORY);
     }
 
     static class JvmIdGetException extends IOException {

--- a/src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java
+++ b/src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java
@@ -81,6 +81,7 @@ import io.cryostat.net.web.WebModule;
 import io.cryostat.net.web.WebServer;
 import io.cryostat.net.web.http.HttpMimeType;
 import io.cryostat.platform.PlatformClient;
+import io.cryostat.recordings.JvmIdHelper.JvmIdGetException;
 import io.cryostat.recordings.RecordingMetadataManager.Metadata;
 import io.cryostat.rules.ArchivePathException;
 import io.cryostat.rules.ArchivedRecordingInfo;
@@ -113,6 +114,7 @@ public class RecordingArchiveHelper {
     public static final String ARCHIVES = "archives";
     public static final String UPLOADED_RECORDINGS_SUBDIRECTORY = "uploads";
     public static final String DEFAULT_CACHED_REPORT_SUBDIRECTORY = "default";
+    public static final String LOST_RECORDINGS_SUBDIRECTORY = "lost";
     private static final String CONNECT_URL = "connectUrl";
 
     RecordingArchiveHelper(
@@ -142,7 +144,7 @@ public class RecordingArchiveHelper {
         this.base32 = base32;
     }
 
-    // on startup migration and jvmId transfer method for archived recordings
+    // startup migration for archived recordings
     protected void migrate(ExecutorService executor) throws Exception {
         List<String> subdirectories = fs.listDirectoryChildren(archivedRecordingsPath);
         List<Future<Void>> futures = new ArrayList<>(subdirectories.size());
@@ -150,18 +152,20 @@ public class RecordingArchiveHelper {
             Future<Void> future =
                     executor.submit(
                             () -> {
+                                Path subdirectoryPath = null;
                                 try {
+                                    if (jvmIdHelper.isSpecialDirectory(subdirectoryName)) {
+                                        logger.info(
+                                                "Skipping archive migration: appears to be a"
+                                                        + " special location: {}",
+                                                subdirectoryName);
+                                        return null;
+                                    }
                                     logger.info(
                                             "Found archived recordings subdirectory: {}",
                                             subdirectoryName);
-                                    // FIXME: refactor structure to remove file-uploads (v1
-                                    // RecordingsPostBodyHandler)
-                                    if (subdirectoryName.equals("file-uploads")
-                                            || subdirectoryName.equals("uploads")) {
-                                        logger.info("Skipping: appears to be an upload location");
-                                        return null;
-                                    }
-                                    Path subdirectoryPath =
+
+                                    subdirectoryPath =
                                             archivedRecordingsPath.resolve(subdirectoryName);
                                     String connectUrl;
                                     try {
@@ -200,7 +204,28 @@ public class RecordingArchiveHelper {
                                         Files.move(oldLocation, newLocation);
                                     }
                                     FileUtils.deleteQuietly(subdirectoryPath.toFile());
-
+                                } catch (JvmIdGetException e) {
+                                    logger.warn(
+                                            "Failed to migrate archived recordings for {} - no"
+                                                    + " target connection available to {}",
+                                            subdirectoryName,
+                                            e.getTarget());
+                                    Path lostPath =
+                                            archivedRecordingsPath.resolve(
+                                                    LOST_RECORDINGS_SUBDIRECTORY);
+                                    if (!fs.exists(lostPath)) {
+                                        fs.createDirectory(lostPath);
+                                    }
+                                    for (String file : fs.listDirectoryChildren(subdirectoryPath)) {
+                                        Path oldLocation = subdirectoryPath.resolve(file);
+                                        Path newLocation = lostPath.resolve(file);
+                                        logger.info(
+                                                "Moving lost recordings: {} -> {}",
+                                                oldLocation,
+                                                newLocation);
+                                        Files.move(oldLocation, newLocation);
+                                    }
+                                    FileUtils.deleteQuietly(subdirectoryPath.toFile());
                                 } catch (IOException e) {
                                     logger.warn(e);
                                 } catch (CancellationException e) {
@@ -241,7 +266,7 @@ public class RecordingArchiveHelper {
             }
 
             logger.info(
-                    "{} Archives subdirectory transfer: {} -> {}",
+                    "[{}]: Archives subdirectory transfer: {} -> {}",
                     connectUrl,
                     subdirectoryPath,
                     newJvmId);
@@ -256,7 +281,7 @@ public class RecordingArchiveHelper {
             fs.deleteIfExists(subdirectoryPath);
 
             logger.info(
-                    "{} Archives subdirectory successfully transferred: {} -> {}",
+                    "[{}]: Archives subdirectory successfully transferred: {} -> {}",
                     connectUrl,
                     subdirectoryPath,
                     newJvmId);
@@ -279,6 +304,8 @@ public class RecordingArchiveHelper {
         } else if (subdirectory.getFileName().toString().equals(UPLOADED_RECORDINGS_SUBDIRECTORY)
                 || subdirectory.getFileName().toString().equals("file-uploads")) {
             future.complete(UPLOADED_RECORDINGS_SUBDIRECTORY);
+        } else if (subdirectory.getFileName().toString().equals(LOST_RECORDINGS_SUBDIRECTORY)) {
+            future.complete(LOST_RECORDINGS_SUBDIRECTORY);
         } else {
             Optional<String> connectUrl = Optional.empty();
             try {
@@ -299,10 +326,7 @@ public class RecordingArchiveHelper {
     }
 
     Path getRecordingSubdirectoryPath(String jvmId) {
-        String subdirectory =
-                jvmId.equals(UPLOADED_RECORDINGS_SUBDIRECTORY)
-                        ? UPLOADED_RECORDINGS_SUBDIRECTORY
-                        : base32.encodeAsString(jvmId.getBytes(StandardCharsets.UTF_8));
+        String subdirectory = jvmIdHelper.jvmIdToSubdirectoryName(jvmId);
         return archivedRecordingsPath.resolve(subdirectory);
     }
 
@@ -384,7 +408,7 @@ public class RecordingArchiveHelper {
             String subdirectoryName, String recordingName) {
         CompletableFuture<ArchivedRecordingInfo> future = new CompletableFuture<>();
         try {
-            String jvmId = new String(base32.decode(subdirectoryName), StandardCharsets.UTF_8);
+            String jvmId = jvmIdHelper.subdirectoryNameToJvmId(subdirectoryName);
             Path subdirectoryPath = archivedRecordingsPath.resolve(subdirectoryName);
             Path recordingPath = subdirectoryPath.resolve(recordingName);
             validateSavePath(recordingName, recordingPath);
@@ -642,7 +666,7 @@ public class RecordingArchiveHelper {
                 }
                 Path subdirectory = archivedRecordingsPath.resolve(subdirectoryName);
                 String targetId = getConnectUrlFromPath(subdirectory).get();
-                String jvmId = new String(base32.decode(subdirectoryName), StandardCharsets.UTF_8);
+                String jvmId = jvmIdHelper.subdirectoryNameToJvmId(subdirectoryName);
                 List<String> files = this.fs.listDirectoryChildren(subdirectory);
                 List<ArchivedRecordingInfo> temp =
                         files.stream()
@@ -744,8 +768,9 @@ public class RecordingArchiveHelper {
 
     public Future<Path> getRecordingPathFromPath(String subdirectoryName, String recordingName) {
         try {
+            boolean checkConnectUrl = !jvmIdHelper.isSpecialDirectory(subdirectoryName);
             Path path = archivedRecordingsPath.resolve(subdirectoryName).resolve(recordingName);
-            validateRecordingPath(Optional.of(path), recordingName, false);
+            validateRecordingPath(Optional.of(path), recordingName, checkConnectUrl);
             return CompletableFuture.completedFuture(path);
         } catch (RecordingNotFoundException | ArchivePathException e) {
             return CompletableFuture.failedFuture(e);
@@ -758,7 +783,7 @@ public class RecordingArchiveHelper {
             List<String> subdirectories = this.fs.listDirectoryChildren(archivedRecordingsPath);
             Optional<Path> optional =
                     searchSubdirectories(subdirectories, archivedRecordingsPath, recordingName);
-            validateRecordingPath(optional, recordingName, true);
+            validateRecordingPath(optional, recordingName, false);
             future.complete(optional.get());
         } catch (RecordingNotFoundException | IOException | ArchivePathException e) {
             future.completeExceptionally(e);
@@ -785,7 +810,7 @@ public class RecordingArchiveHelper {
             if (archivedRecording == null) {
                 throw new RecordingNotFoundException(sourceTarget, recordingName);
             }
-            validateRecordingPath(Optional.of(archivedRecording), recordingName, false);
+            validateRecordingPath(Optional.of(archivedRecording), recordingName, true);
             future.complete(archivedRecording);
         } catch (RecordingNotFoundException | ArchivePathException | IOException e) {
             future.completeExceptionally(e);
@@ -840,7 +865,7 @@ public class RecordingArchiveHelper {
     }
 
     private void validateRecordingPath(
-            Optional<Path> optional, String recordingName, boolean isUploads)
+            Optional<Path> optional, String recordingName, boolean checkConnectUrl)
             throws RecordingNotFoundException, ArchivePathException {
         if (optional.isEmpty()) {
             throw new RecordingNotFoundException(ARCHIVES, recordingName);
@@ -855,7 +880,7 @@ public class RecordingArchiveHelper {
         if (!fs.isReadable(archivedRecording)) {
             throw new ArchivePathException(archivedRecording.toString(), "is not readable");
         }
-        if (!isUploads && !fs.exists(archivedRecording.resolveSibling(CONNECT_URL))) {
+        if (checkConnectUrl && !fs.exists(archivedRecording.resolveSibling(CONNECT_URL))) {
             throw new ArchivePathException(
                     archivedRecording.resolveSibling(CONNECT_URL).toString(), "does not exist");
         }

--- a/src/main/java/io/cryostat/recordings/RecordingMetadataManager.java
+++ b/src/main/java/io/cryostat/recordings/RecordingMetadataManager.java
@@ -151,17 +151,23 @@ public class RecordingMetadataManager extends AbstractVerticle
         RecordingArchiveHelper archiveHelper = archiveHelperProvider.get();
         try {
             this.fs.listDirectoryChildren(recordingMetadataDir).stream()
-                    .peek(
-                            n ->
-                                    logger.info(
-                                            "Peeking contents of recordingMetadata directory: {}",
-                                            n))
+                    .peek(n -> logger.info("Peeking contents of metadata directory: {}", n))
                     .map(recordingMetadataDir::resolve)
                     .forEach(
                             subdirectory -> {
                                 if (fs.isDirectory(subdirectory)) {
                                     try {
-                                        if (this.fs.listDirectoryChildren(subdirectory).isEmpty()) {
+                                        String subdirectoryName =
+                                                subdirectory.getFileName().toString();
+                                        if (jvmIdHelper.isSpecialDirectory(subdirectoryName)) {
+                                            logger.info(
+                                                    "Skipping metadata validation: appears to be a"
+                                                            + " special location: {}",
+                                                    subdirectoryName);
+                                            return;
+                                        } else if (this.fs
+                                                .listDirectoryChildren(subdirectory)
+                                                .isEmpty()) {
                                             logger.info(
                                                     "Deleting empty recording metadata directory:"
                                                             + " {}",
@@ -357,7 +363,7 @@ public class RecordingMetadataManager extends AbstractVerticle
                                     }
                                 } else {
                                     logger.warn(
-                                            "Recording Metadata subdirectory {} is"
+                                            "Recording metadata subdirectory {} is"
                                                     + " neither a directory nor a file",
                                             subdirectory);
                                     // invalid
@@ -369,7 +375,7 @@ public class RecordingMetadataManager extends AbstractVerticle
             future.complete();
         } catch (IOException e) {
             logger.error(
-                    "Could not read recordingMetadataDirectory! {}, msg: {}",
+                    "Could not read recording metadata directory! {}, msg: {}",
                     recordingMetadataDir,
                     e.getMessage());
             future.fail(e.getCause());
@@ -515,7 +521,8 @@ public class RecordingMetadataManager extends AbstractVerticle
                         .get()
                         .getConnectUrlFromPath(archivedRecordingsPath.resolve(subdirectoryName))
                         .get();
-        String jvmId = new String(base32.decode(subdirectoryName), StandardCharsets.UTF_8);
+        String jvmId = jvmIdHelper.subdirectoryNameToJvmId(subdirectoryName);
+
         Path metadataPath = this.getMetadataPath(jvmId, recordingName);
         fs.writeString(
                 metadataPath,
@@ -837,10 +844,7 @@ public class RecordingMetadataManager extends AbstractVerticle
     }
 
     private Path getMetadataPath(String jvmId) throws IOException {
-        String subdirectory =
-                jvmId.equals(UPLOADS)
-                        ? UPLOADS
-                        : base32.encodeAsString(jvmId.getBytes(StandardCharsets.UTF_8));
+        String subdirectory = jvmIdHelper.jvmIdToSubdirectoryName(jvmId);
 
         Path parentDir = recordingMetadataDir.resolve(subdirectory);
         if (parentDir == null) {

--- a/src/main/java/io/cryostat/recordings/RecordingsModule.java
+++ b/src/main/java/io/cryostat/recordings/RecordingsModule.java
@@ -216,6 +216,7 @@ public abstract class RecordingsModule {
             @Named(Variables.JMX_CONNECTION_TIMEOUT) long connectionTimeoutSeconds,
             CredentialsManager credentialsManager,
             DiscoveryStorage storage,
+            Base32 base32,
             Logger logger) {
         return new JvmIdHelper(
                 targetConnectionManager,
@@ -224,6 +225,7 @@ public abstract class RecordingsModule {
                 connectionTimeoutSeconds,
                 ForkJoinPool.commonPool(),
                 Scheduler.systemScheduler(),
+                base32,
                 logger);
     }
 }

--- a/src/test/java/io/cryostat/recordings/RecordingArchiveHelperTest.java
+++ b/src/test/java/io/cryostat/recordings/RecordingArchiveHelperTest.java
@@ -131,6 +131,25 @@ class RecordingArchiveHelperTest {
                 .thenReturn(notificationBuilder);
         lenient().when(notificationBuilder.message(Mockito.any())).thenReturn(notificationBuilder);
         lenient().when(notificationBuilder.build()).thenReturn(notification);
+        lenient()
+                .when(jvmIdHelper.jvmIdToSubdirectoryName(Mockito.anyString()))
+                .thenAnswer(
+                        new Answer<String>() {
+                            @Override
+                            public String answer(InvocationOnMock invocation) throws Throwable {
+                                return invocation.getArgument(0);
+                            }
+                        });
+        lenient()
+                .when(jvmIdHelper.subdirectoryNameToJvmId(Mockito.anyString()))
+                .thenAnswer(
+                        new Answer<String>() {
+                            @Override
+                            public String answer(InvocationOnMock invocation) throws Throwable {
+                                return invocation.getArgument(0);
+                            }
+                        });
+
         this.recordingArchiveHelper =
                 new RecordingArchiveHelper(
                         fs,
@@ -226,7 +245,6 @@ class RecordingArchiveHelperTest {
         Mockito.when(connection.getJMXURL())
                 .thenReturn(
                         (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9092/jmxrmi")));
-        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
 
         Instant now = Instant.now();
         Mockito.when(clock.now()).thenReturn(now);
@@ -304,7 +322,6 @@ class RecordingArchiveHelperTest {
         Mockito.when(connection.getJMXURL())
                 .thenReturn(
                         (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9091/jmxrmi")));
-        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
 
         Instant now = Instant.now();
         Mockito.when(clock.now()).thenReturn(now);
@@ -399,7 +416,6 @@ class RecordingArchiveHelperTest {
         Mockito.when(connection.getJMXURL())
                 .thenReturn(
                         (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9092/jmxrmi")));
-        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
         Mockito.when(connection.getHost()).thenReturn("some-hostname.local");
 
         Instant now = Instant.now();
@@ -479,7 +495,6 @@ class RecordingArchiveHelperTest {
         Mockito.when(connection.getJMXURL())
                 .thenReturn(
                         (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9092/jmxrmi")));
-        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
         Mockito.when(connection.getHost()).thenReturn("some-hostname.local");
 
         Instant now = Instant.now();
@@ -566,7 +581,6 @@ class RecordingArchiveHelperTest {
         Mockito.when(connection.getJMXURL())
                 .thenReturn(
                         (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9092/jmxrmi")));
-        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
 
         Instant now = Instant.now();
         Mockito.when(clock.now()).thenReturn(now);
@@ -636,7 +650,6 @@ class RecordingArchiveHelperTest {
         Mockito.when(connection.getJMXURL())
                 .thenReturn(
                         (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9092/jmxrmi")));
-        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
 
         Instant now = Instant.now();
         Mockito.when(clock.now()).thenReturn(now);
@@ -703,7 +716,6 @@ class RecordingArchiveHelperTest {
         Mockito.when(connection.getJMXURL())
                 .thenReturn(
                         (new JMXServiceURL("service:jmx:rmi:///jndi/rmi://cryostat:9092/jmxrmi")));
-        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn("encodedServiceUri");
 
         Instant now = Instant.now();
         Mockito.when(clock.now()).thenReturn(now);
@@ -985,7 +997,6 @@ class RecordingArchiveHelperTest {
                 .when(jvmIdHelper.getJvmId(Mockito.eq(targetIdUploads)))
                 .thenReturn(targetIdUploads);
         lenient().when(jvmIdHelper.getJvmId(Mockito.eq(targetIdTarget))).thenReturn(targetIdTarget);
-        Mockito.when(base32.encodeAsString(Mockito.any())).thenReturn(targetIdTarget);
         Path specificRecordingsPath = Path.of("/some/path/");
         Mockito.when(archivedRecordingsPath.resolve(Mockito.anyString()))
                 .thenReturn(specificRecordingsPath);
@@ -1013,6 +1024,15 @@ class RecordingArchiveHelperTest {
 
         Mockito.when(recordingMetadataManager.getMetadata(Mockito.any(), Mockito.anyString()))
                 .thenReturn(new Metadata());
+
+        Mockito.when(jvmIdHelper.jvmIdToSubdirectoryName(Mockito.anyString()))
+                .thenAnswer(
+                        new Answer<String>() {
+                            @Override
+                            public String answer(InvocationOnMock invocation) throws Throwable {
+                                return invocation.getArgument(0);
+                            }
+                        });
 
         // Test get recordings from uploads
         List<ArchivedRecordingInfo> result =
@@ -1064,10 +1084,6 @@ class RecordingArchiveHelperTest {
                 .thenReturn(Path.of(subdirectories.get(1)));
         Mockito.when(fs.listDirectoryChildren(Path.of(subdirectories.get(1))))
                 .thenReturn(List.of("123recording", "connectUrl"));
-
-        Mockito.when(base32.decode(Mockito.anyString()))
-                .thenReturn("encodedJvmIdA".getBytes())
-                .thenReturn("encodedJvmId123".getBytes());
 
         BufferedReader reader = Mockito.mock(BufferedReader.class);
         Mockito.when(fs.readFile(Mockito.any(Path.class))).thenReturn(reader);

--- a/src/test/java/io/cryostat/recordings/RecordingMetadataManagerTest.java
+++ b/src/test/java/io/cryostat/recordings/RecordingMetadataManagerTest.java
@@ -79,7 +79,9 @@ import org.junit.jupiter.params.provider.ValueSource;
 import org.mockito.InOrder;
 import org.mockito.Mock;
 import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
 import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.stubbing.Answer;
 
 @ExtendWith(MockitoExtension.class)
 public class RecordingMetadataManagerTest {
@@ -121,6 +123,15 @@ public class RecordingMetadataManagerTest {
                 .thenReturn(notificationBuilder);
         lenient().when(notificationBuilder.message(Mockito.any())).thenReturn(notificationBuilder);
         lenient().when(notificationBuilder.build()).thenReturn(notification);
+        lenient()
+                .when(jvmIdHelper.jvmIdToSubdirectoryName(Mockito.anyString()))
+                .thenAnswer(
+                        new Answer<String>() {
+                            @Override
+                            public String answer(InvocationOnMock invocation) throws Throwable {
+                                return invocation.getArgument(0);
+                            }
+                        });
 
         this.recordingMetadataManager =
                 new RecordingMetadataManager(

